src/_P014_SI70xx.ino

```diff
@@ -0,0 +1,198 @@
+      break;
+    }
+
+    case PLUGIN_I2C_HAS_ADDRESS:
+    case PLUGIN_WEBFORM_SHOW_I2C_PARAMS:
+    {
+      const uint8_t i2cAddressValues[] = { SI70xx_I2C_ADDRESS, SI7013_I2C_ADDRESS_AD0_1 };
+
+      if (function == PLUGIN_WEBFORM_SHOW_I2C_PARAMS) {
+        addFormSelectorI2C(F("i2c_addr"), 2, i2cAddressValues, P014_I2C_ADDRESS);
+        addFormNote(F("ADO Low=0x40, High=0x41"));
+      } else {
+        success = intArrayContains(2, i2cAddressValues, event->Par1);
+      }
+
+      break;
+    }
+
+
+    case PLUGIN_WEBFORM_LOAD:
+    {
+        # define P014_RESOLUTION_OPTIONS 4
+
+      uint8_t choice = P014_RESOLUTION;
+      const __FlashStringHelper *options[P014_RESOLUTION_OPTIONS];
+      int optionValues[P014_RESOLUTION_OPTIONS];
+      optionValues[0] = SI70xx_RESOLUTION_14T_12RH;
+      options[0]      = F("Temp 14 bits / RH 12 bits");
+      optionValues[1] = SI70xx_RESOLUTION_13T_10RH;
+      options[1]      = F("Temp 13 bits / RH 10 bits");
+      optionValues[2] = SI70xx_RESOLUTION_12T_08RH;
+      options[2]      = F("Temp 12 bits / RH  8 bits");
+      optionValues[3] = SI70xx_RESOLUTION_11T_11RH;
+      options[3]      = F("Temp 11 bits / RH 11 bits");
+      addFormSelector(F("Resolution"), F("p014_res"), P014_RESOLUTION_OPTIONS, options, optionValues, choice);
+
+      addFormNumericBox("ADC Filter Power", F("p014_filter"), P014_FILTER_POWER, 0, 4);
+
+      // addUnit(F("bits"));
+
+      success = true;
+      break;
+    }
+
+    case PLUGIN_WEBFORM_SAVE:
+    {
+      P014_I2C_ADDRESS  = getFormItemInt(F("i2c_addr"));
+      P014_RESOLUTION   = getFormItemInt(F("p014_res"));
+      P014_FILTER_POWER = getFormItemInt(F("p014_filter"));
+        
+      // Force device setup next time
+      P014_data_struct *P014_data = static_cast<P014_data_struct *>(getPluginTaskData(event->TaskIndex));
+      if (nullptr != P014_data) {
+        P014_data->state = P014_state::Uninitialized;
+      }
+
+      success = true;
+      break;
+    }
+
+
+    case PLUGIN_INIT:
+    {
+      initPluginTaskData(event->TaskIndex, new (std::nothrow) P014_data_struct());
+      P014_data_struct *P014_data = static_cast<P014_data_struct *>(getPluginTaskData(event->TaskIndex));
+
+      if (nullptr == P014_data) {
+        return success;
+      }
+      if (P014_data->init(P014_I2C_ADDRESS, P014_RESOLUTION)) {
+        success = true;
+      }else{
+        UserVar[event->BaseVarIndex]     = NAN;
+        UserVar[event->BaseVarIndex + 1] = NAN;
+        UserVar[event->BaseVarIndex + 2] = NAN;
+      }
+      break;
+    }
+
+
+    case PLUGIN_READ:
+    {
+      P014_data_struct *P014_data = static_cast<P014_data_struct *>(getPluginTaskData(event->TaskIndex));
+
+      if (nullptr != P014_data) {
+        if (P014_data->state == P014_state::Error){
+          UserVar[event->BaseVarIndex]     = NAN;
+          UserVar[event->BaseVarIndex + 1] = NAN;
+          UserVar[event->BaseVarIndex + 2] = NAN;
```

Comment:
This is also a functional change from older setups.
What is the reason to set these to NaN?
Not all controllers do handle NaN values very well, thus this change may break existing node configurations.

Reply:
Maybe they should as the type is float so NAN is a valid response.

Reply:
Some controllers just check to see if there is any invalid value and refuse to send if there is.
Not sure if that's a wise decision, but that's how it is right now.

We are working on making a generic selection to be used in any plugin to let the user select what a "failed" value should be.
You can see it being implemented in the BME280 plugin.